### PR TITLE
Revert "drivers: move eeprom_slave driver to tests directory"

### LINKED ISSUE #30557: i2c slave driver removed
@pabigot 
Reviewing the 2.4 release notes I noticed issue  #27303 removed the i2c slave driver.  This was the only i2c slave driver available in the tree and its removal without a replacement is particularly troubling since I use this driver.  Please put it back.

### PR DESCRIPTION
**Pull request by pabigot**
This reverts commit cabbd916cfba20f39a91922b740d914f38a21b2d.

This functionality is considered to be useful enough that it should be restored as a stable Zephyr API.

Closes #30557

### CODE REVIEW

**Comment by pabigot** [comment]
Any volunteers to own this thing?  Needs a CODEOWNER entry.  Maybe @brix since it's pretending to be an EEPROM?

**Comment by pabigot** [comment]
@sslupsky please verify this.

**Comment by henrikbrixandersen** [comment]
> Any volunteers to own this thing? Needs a CODEOWNER entry. Maybe @henrikbrixandersen since it's pretending to be an EEPROM?

@pabigot Sure. You can put me as code owner.

**Review by henrikbrixandersen — APPROVED** [review]

**Comment by pabigot** [comment]
@henrikbrixandersen thanks, updated.

File: CODEOWNERS [hunk 1]
```diff
@@ -208,6 +208,7 @@
 /drivers/i2c/i2c_emul.c                   @sjg20
 /drivers/i2c/i2c_shell.c                  @nashif
 /drivers/i2c/Kconfig.i2c_emul             @sjg20
+/drivers/i2c/slave/                       @henrikbrixandersen
```
Comment (henrikbrixandersen) [inline CODEOWNERS:211]: This may become more than I signed up for in the future 😉 Can we limit it to just this one driver file?

Comment (pabigot) [reply CODEOWNERS:211]: There's nothing else in there, but yeah, I can limit it to whatever the check complains about.

**Review by nashif — APPROVED** [review]

**Pull request merged**